csharp/src/SeedLang/Interpreter/Compiler.cs

```diff
@@ -96,12 +78,64 @@ private struct ExpressionInfo {
     }
 
     protected override void Visit(ComparisonExpression comparison) {
-      throw new NotImplementedException();
+      // TODO: support multiple operators. Current implementation only supports two operators.
+      bool needFirstRegister = !TryGetRegisterOrConstantId(comparison.First, out uint firstId);
+      uint first = needFirstRegister ? _registerAllocator.AllocateTempVariable() : firstId;
+      if (needFirstRegister) {
+        _registerForSubExpr = first;
+        Visit(comparison.First);
+      }
+      bool needSecondRegister = !TryGetRegisterOrConstantId(comparison.Exprs[0], out uint secondId);
+      uint second = needSecondRegister ? _registerAllocator.AllocateTempVariable() : secondId;
+      if (needSecondRegister) {
+        _registerForSubExpr = second;
+        Visit(comparison.Exprs[0]);
+      }
+      Opcode op = Opcode.EQ;
+      bool expectedResult = false;
+      switch (comparison.Ops[0]) {
+        case ComparisonOperator.Less:
+          op = Opcode.LT;
+          expectedResult = true;
+          break;
+        case ComparisonOperator.Greater:
+          op = Opcode.LE;
+          expectedResult = false;
+          break;
+        case ComparisonOperator.LessEqual:
+          op = Opcode.LE;
+          expectedResult = true;
+          break;
+        case ComparisonOperator.GreaterEqual:
+          op = Opcode.LT;
+          expectedResult = false;
+          break;
+        case ComparisonOperator.EqEqual:
+          op = Opcode.EQ;
+          expectedResult = true;
+          break;
+        case ComparisonOperator.NotEqual:
+          op = Opcode.EQ;
+          expectedResult = false;
+          break;
+      }
+      _chunk.Emit(op, expectedResult ? 1u : 0u, first, second, comparison.Range);
+      if (needFirstRegister) {
+        _registerAllocator.DeallocateVariable();
```

Comment:
There are many `if (condition) DeallocateVariable()` code blocks. Could there be resource leak risks if those kind of conditions are not handled carefully? ... Just to make sure that you have already considered different solutions and still think this "allocate-deallocate" pair is the best fit.

Reply:
Added expression scope concept in RegisterAllocator to simplify client logic. Allocated temporary variables will be deallocated automatically when exiting expression scopes.

Reply:
This is cool!